File: src/java/magmac/api/collect/TupleCollector.java
package magmac.api.collect;

import magmac.api.Tuple2;
import magmac.api.iter.collect.Collector;

public record TupleCollector<A, AC, B, BC>(
        Collector<A, AC> leftCollector,
        Collector<B, BC> rightCollector
) implements Collector<Tuple2<A, B>, Tuple2<AC, BC>> {
    @Override
    public Tuple2<AC, BC> createInitial() {
        return new Tuple2<>(this.leftCollector.createInitial(), this.rightCollector.createInitial());
    }

    @Override
    public Tuple2<AC, BC> fold(Tuple2<AC, BC> current, Tuple2<A, B> element) {
        AC leftValue = this.leftCollector.fold(current.left(), element.left());
        BC rightValue = this.rightCollector.fold(current.right(), element.right());
        return new Tuple2<>(leftValue, rightValue);
    }
}

File: src/web/magmac/api/collect/TupleCollector.ts
import { Tuple2 } from "../../../magmac/api/Tuple2";
import { Collector } from "../../../magmac/api/iter/collect/Collector";
export class TupleCollector<A,  AC,  B,  BC> {
	createInitial() : Tuple2<AC, BC> {return new Tuple2<>( this.leftCollector.createInitial( ), this.rightCollector.createInitial( ));;}
	fold(current : Tuple2<AC, BC>, element : Tuple2<A, B>) : Tuple2<AC, BC> {break;break;return new Tuple2<>( leftValue, rightValue);;}
}
